### PR TITLE
Add test for empty product list branch in resources.ts

### LINKED ISSUE #85: codecov/project failed on #49 due to ESM migration changing coverage measurement
## Problem

After merging PR #49 (commit 61f93dc), the `codecov/project` check failed:

> **codecov/project**: failure — "97.36% (-0.64%) compared to aa9a6b0"

All GitHub Actions (`make`, `eslint`, `titles`, etc.) passed, including unit tests (21/21).

## Root Cause

The failure was caused by switching the Jest preset in `jest.config.ts`:

```diff
- preset: 'ts-jest'
+ preset: 'ts-jest/presets/default-esm'
```

This was part of the ESM migration (#44 / #49), which also added `"type": "module"` to `package.json`, `.js` suffixes in imports, and `moduleNameMapper` for Jest.

`ts-jest/presets/default-esm` enables `useESM: true`, which changes how code is instrumented for coverage. The coverage calculation for branch coverage shifted, causing the uncovered branch in `src/resources.ts:24` (`if (csv.length !== 0)`) to be detected differently, resulting in a 0.64% drop.

## Impact on Future PRs

This is a **one-time issue** — it only occurred because Codecov compared this commit against its parent (`aa9a6b0`), which used a different Jest preset. Subsequent PRs will use this commit as their baseline, so the fluctuation won't repeat.

## Recommendations

1. **Add `codecov.yml`** with a sensitivity threshold (e.g. `threshold: 1` for `project`) so that instrumentation-related coverage fluctuations don't fail CI in the future.

2. **Add a test for empty resource list** in `src/resources.ts:24`. Branch coverage for this `if` is currently 50% (only non-empty CSV is tested). A test where `baza` returns `''` would cover the second branch and bring branch coverage to 100%.

3. **Check `tools/call` params in `tools.test.ts`** — during the ESM refactoring the params format was changed:
   ```diff
   - arguments: { product, concern }
   + product, concern  // directly in params
   ```
   The MCP protocol expects `{ name, arguments }`. The current test passes, but relies on the SDK not validating this strictly. Better to restore `arguments`.

## Jest Context

This also illustrates that `ts-jest/presets/default-esm` remains experimental (as warned by `--experimental-vm-modules`). Coverage in ESM mode can differ from CJS, which should be considered when configuring CI.

### PR DESCRIPTION
**Pull request by MaksZhukov**
## Summary

- Adds a test covering the case where `baza('/products', 'GET', ...)` returns an empty string, exercising the `else` side of the `if (csv.length !== 0)` branch in `src/resources.ts`.
- Restores branch coverage on `src/resources.ts` from 50% to 100%, fixing the `codecov/project` regression described in the issue (root cause: the ESM Jest preset switch changed how this untested branch was counted).

## Why

Issue #85 identified that `src/resources.ts:24` had only 50% branch coverage — the non-empty CSV path was tested, but the empty-string path (a product-less `baza` response) was not. This is recommendation #2 from the issue.

## Test plan

- [x] `make test` — all 22 tests pass (was 21; added 1)
- [x] Coverage report shows `resources.ts` at 100% branches/statements/functions/lines (was 50% branches)
- [x] `npx eslint . --config eslint.config.mjs` — no errors
- [x] `npx tsc --noEmit --project tsconfig.json` — no errors

Closes #85

### CODE REVIEW

File: test/resources.test.ts [hunk 1]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
     expect(content).toHaveProperty('text');
   });
 
+  test('lists no resources when baza returns an empty product list', async (): Promise<void> => {
+    mock.mockImplementation(async (path, method, params, body) => {
+      if (path === '/products' && method === 'GET') {
+        return '';
```
Comment (Favixx) [inline test/resources.test.ts:84]: This is the input that shows the branch is redundant: with `csv = ''`, the code inside the `if` produces `[]`, which is what `list` was initialised to. So this test passes identically whether the `if (csv.length !== 0)` guard is present or removed — it cannot fail for the reason it exists. A test that pins behaviour rather than a branch would be one where the two differ, and here there is no such input.

File: test/resources.test.ts [hunk 2]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
...
+      }
+      if (path === '/mcp/resource' && method === 'PUT' && params.name === 'product') {
```
Comment (Favixx) [inline test/resources.test.ts:86]: This arm of the mock can never fire in this test: with no products in the list, nothing ever reads an individual resource, so `/mcp/resource` is not requested. It is copied from the neighbouring test where it is needed. Dropping it would make the mock say what this test actually depends on — that `/products` returns empty — which is the whole point of the case.

File: test/resources.test.ts [hunk 3]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
     expect(content).toHaveProperty('text');
   });
 
+  test('lists no resources when baza returns an empty product list', async (): Promise<void> => {
+    mock.mockImplementation(async (path, method, params, body) => {
```
Comment (Favixx) [inline test/resources.test.ts:82]: This mock is the third near-identical copy in the file — same three arms, same fallback, differing only in what `/products` returns. A small helper taking the products string and returning the implementation would collapse all three and make each test one line of setup plus its assertions, which is where the actual difference between them lives.

File: test/resources.test.ts [hunk 4]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
...
+      return body;
+    });
+    const answer = await once({
+      jsonrpc: '2.0' as const,
+      id: 1,
+      method: 'resources/list'
+    });
+    expect(answer).toHaveProperty('result');
+    expect(answer.result).toHaveProperty('resources');
+    expect(answer.result?.resources).toEqual([]);
```
Comment (Favixx) [inline test/resources.test.ts:98]: Good assertion — `toEqual([])` is precise where `toHaveLength(0)` would also pass for a non-array with a length. Worth noting it is also the assertion the test above already makes for a different input, which is the clearest sign that the two cases converge on the same code path.

File: test/resources.test.ts [hunk 5]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
     expect(content).toHaveProperty('text');
   });
 
+  test('lists no resources when baza returns an empty product list', async (): Promise<void> => {
```
Comment (Favixx) [inline test/resources.test.ts:81]: The name describes the input rather than the guarantee. What a reader wants to know is that an empty response from `baza` produces a valid, empty MCP `resources/list` reply rather than an error or a malformed result — that is the behaviour worth protecting, and it stays true regardless of whether the `if` survives a later cleanup.

File: test/resources.test.ts [hunk 6]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
...
+    mock.mockImplementation(async (path, method, params, body) => {
+      if (path === '/products' && method === 'GET') {
+        return '';
+      }
+      if (path === '/mcp/resource' && method === 'PUT' && params.name === 'product') {
+        return `Details for product ${params.product}`;
+      }
+      return body;
```
Comment (Favixx) [inline test/resources.test.ts:89]: The fallback returns the request body, which means an unexpected call silently succeeds with an echo instead of failing. In a test whose subject is "nothing should be requested beyond `/products`", making the fallback throw would turn an unexpected call into a clear failure — and would have surfaced the unused `/mcp/resource` arm above.

File: test/resources.test.ts [hunk 7]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
...
+    });
+    const answer = await once({
+      jsonrpc: '2.0' as const,
+      id: 1,
+      method: 'resources/list'
+    });
+    expect(answer).toHaveProperty('result');
```
Comment (Favixx) [inline test/resources.test.ts:96]: These two assertions are worth keeping — they check the response is a well-formed JSON-RPC result rather than an error object, which is the part that would break if an empty CSV caused a throw somewhere upstream. That is the genuine risk this test guards against, and it is independent of the branch question.

File: test/resources.test.ts [hunk 8]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
...
+    mock.mockImplementation(async (path, method, params, body) => {
+      if (path === '/products' && method === 'GET') {
+        return '';
+      }
+      if (path === '/mcp/resource' && method === 'PUT' && params.name === 'product') {
+        return `Details for product ${params.product}`;
+      }
+      return body;
+    });
+    const answer = await once({
```
Comment (Favixx) [inline test/resources.test.ts:91]: Driving this through the real JSON-RPC entry point rather than calling `list()` directly is the right choice: it covers the `ResourceTemplate` wiring as well as the handler. It also means the test would catch a regression in how an empty `resources` array is serialised, which a unit-level call would not.

File: test/resources.test.ts [hunk 9]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
...
+      jsonrpc: '2.0' as const,
+      id: 1,
+      method: 'resources/list'
+    });
+    expect(answer).toHaveProperty('result');
+    expect(answer.result).toHaveProperty('resources');
+    expect(answer.result?.resources).toEqual([]);
+  });
```
Comment (Favixx) [inline test/resources.test.ts:99]: One case adjacent to this one is still uncovered and is not redundant: `baza` rejecting. If `/products` fails, the rejection propagates out of `list()` and the JSON-RPC layer has to turn it into something — an error result, presumably. That path has real behaviour to pin, unlike the empty-string branch, and it is the same three lines of setup.

File: test/resources.test.ts [hunk 10]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
     expect(content).toHaveProperty('text');
   });
 
+  test('lists no resources when baza returns an empty product list', async (): Promise<void> => {
+    mock.mockImplementation(async (path, method, params, body) => {
+      if (path === '/products' && method === 'GET') {
+        return '';
+      }
+      if (path === '/mcp/resource' && method === 'PUT' && params.name === 'product') {
+        return `Details for product ${params.product}`;
```
Comment (Favixx) [inline test/resources.test.ts:87]: Since this arm never runs here, the template string is also never evaluated — so if `params.product` were renamed, this test would keep passing while the neighbouring one broke. Copied setup that cannot execute is the kind that drifts out of step with the code it imitates without anyone noticing.

File: test/resources.test.ts [hunk 11]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
...
+      }
+      return body;
+    });
+    const answer = await once({
+      jsonrpc: '2.0' as const,
+      id: 1,
+      method: 'resources/list'
+    });
+    expect(answer).toHaveProperty('result');
+    expect(answer.result).toHaveProperty('resources');
```
Comment (Favixx) [inline test/resources.test.ts:97]: Worth pairing with a check that nothing else appears alongside it — the MCP list response for an empty set should be `{ resources: [] }` and not, say, carry a `nextCursor`. Asserting the whole object with `toEqual({ resources: [] })` would pin the shape rather than the presence of one key, at no extra cost.

File: test/resources.test.ts [hunk 12]
```diff
@@ -78,6 +78,26 @@ describe('resources', () => {
     expect(content).toHaveProperty('text');
   });
 
+  test('lists no resources when baza returns an empty product list', async (): Promise<void> => {
+    mock.mockImplementation(async (path, method, params, body) => {
+      if (path === '/products' && method === 'GET') {
```
Comment (Favixx) [inline test/resources.test.ts:83]: The description explains that the coverage regression came from the ESM Jest preset changing how this branch was counted, which is a useful and honest root cause. It would be worth one more sentence there recording that the branch was uncovered before the preset change too — otherwise a reader may conclude the preset introduced a problem, when it revealed one.

**Review by Favixx — COMMENTED** [review]
The diagnosis in #85 is right — that branch really was uncovered, and the ESM preset change only made an existing gap visible rather than creating one. The test itself is clean and asserts the correct thing.

My concern is what it locks in. The branch being covered cannot change the outcome. In `src/resources.ts`:

```ts
const csv = await baza('/products', 'GET', {}, '');
let list: Array<Resource> = [];
if (csv.length !== 0) {
  const products = csv.split("\n").filter((p) => p.length > 0);
  list = products.map(...);
}
return { resources: list };
```

When `csv` is `''`, `''.split("\n")` yields `['']`, the `filter` drops it, and `map` returns `[]` — exactly the value `list` already holds. So deleting the `if` and its braces produces identical behaviour for every possible input, including the empty string this test supplies. The guard is dead, and the filter one line down is what actually handles the empty case — which the neighbouring `filters empty products from trailing newlines` test already exercises.

That makes this a choice between two ways of getting to 100%: add a test that pins a redundant branch, or remove the branch and get there with the tests that already exist. The second leaves less code and the same coverage; the first makes the dead `if` harder to remove later, because now something depends on it.

If the branch stays for readability, the test is fine as written — but it is worth saying so in the description, since a future reader will otherwise assume the guard is load-bearing precisely because it has a test.
